File: backend/app/models/recruitment.py
"""recruitments 招募表（含 A2 授权、审核、过期与下架元数据）。"""

from sqlalchemy import JSON, Boolean, Column, Date, DateTime, String, Text, func

from app.db.base import Base
from app.models.match_record import _uuid


class Recruitment(Base):
    __tablename__ = "recruitments"

    recruit_id = Column(String(36), primary_key=True, default=lambda: str(_uuid()))
    publisher_id = Column(String(64), nullable=False, index=True)
    publisher_type = Column(String(20))  # advisor / senior
    type = Column(String(20))  # 实习 / 科研助理 / 招生
    title = Column(String(200))
    req = Column(Text)
    major = Column(String(100))
    deadline = Column(Date)
    is_urgent = Column(Boolean, default=False)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now())
    verified_at = Column(DateTime(timezone=True), nullable=True)
    expires_at = Column(DateTime(timezone=True), nullable=True)
    takedown_at = Column(DateTime(timezone=True), nullable=True)
    review_status = Column(String(20), nullable=False, default="pending_review")
    publication_status = Column(String(20), nullable=False, default="restricted")
    authorization_basis = Column(String(40), nullable=False, default="none")
    consent_id = Column(String(100), nullable=True)
    provenance = Column(JSON, nullable=False, default=dict)
    governance = Column(JSON, nullable=False, default=dict)
    quarantined_fields = Column(JSON, nullable=False, default=dict)
    # —— 立体化扩展（全部可空，向后兼容；迁移 0010）——
    location = Column(String(60), nullable=True)  # 工作地点
    quota = Column(String(20), nullable=True)  # 名额
    compensation = Column(String(60), nullable=True)  # 待遇说明
    duration = Column(String(40), nullable=True)  # 周期
    apply_method = Column(String(200), nullable=True)  # 投递方式（禁联系方式直发）
    tags = Column(JSON, nullable=True)  # 标签列表，最多 10 个
    advisor_id = Column(String(20), nullable=True, index=True)  # 关联导师

File: frontend/src/api/recruitment.ts
import { get, patch, post, remove } from './request'
import type { RecruitmentItem } from '@/types/api'

export function fetchRecruitments(filters?: {
  urgent?: boolean
  tag?: string
  advisor_id?: string
}) {
  const params: Record<string, unknown> = {}
  if (filters?.urgent !== undefined) params.urgent = filters.urgent
  if (filters?.tag) params.tag = filters.tag
  if (filters?.advisor_id) params.advisor_id = filters.advisor_id
  return get<{ data: RecruitmentItem[] }>('/api/recruitments', params)
}

/** 招募详情（公开口径；不存在/未公开返回 404） */
export interface RecruitmentDetail extends RecruitmentItem {
  created_at?: string | null
  verified_at?: string | null
  advisor: { advisor_id: string; name: string; dept: string } | null
  related: RecruitmentItem[]
}

export function fetchRecruitmentDetail(recruitId: string) {
  return get<{ data: RecruitmentDetail }>(`/api/recruitments/${recruitId}`)
}

export interface RecruitmentFormData {
  type: string
  title: string
  req: string
  major: string
  deadline: string
  is_urgent: boolean
  // 立体化扩展（全部选填）
  location?: string
  quota?: string
  compensation?: string
  duration?: string
  apply_method?: string
  tags?: string[]
  advisor_id?: string
}

export interface RecruitmentMutationResult {
  recruit_id: string
  status: 'pending_review'
  publication_status: 'restricted'
  updated?: boolean
}

export function publishRecruitment(
  req: RecruitmentFormData,
  idempotencyKey: string,
) {
  return post<RecruitmentMutationResult>('/api/recruitments', req, {
    headers: { 'Idempotency-Key': idempotencyKey },
  })
}

export interface MyRecruitment extends RecruitmentFormData {
  recruit_id: string
  review_status: string
  publication_status: string
  review_reason?: string | null
  created_at?: string | null
  updated_at?: string | null
}

/** 编辑回填的最小结构（学生端 MyRecruitment 与导师端 MentorRecruitmentItem 均满足） */
export interface EditableRecruitment {
  recruit_id: string
  type: string
  title: string
  req: string
  major: string
  deadline?: string | null
  is_urgent: boolean
  location?: string | null
  quota?: string | null
  compensation?: string | null
  duration?: string | null
  apply_method?: string | null
  tags?: string[] | null
  advisor_id?: string | null
}

export function fetchMyRecruitments() {
  return get<{ data: MyRecruitment[] }>('/api/recruitments/mine')
}

export function updateRecruitment(
  recruitId: string,
  req: RecruitmentFormData,
  idempotencyKey: string,
) {
  return patch<RecruitmentMutationResult>(
    `/api/recruitments/${recruitId}`,
    { ...req, submit_for_review: true },
    { headers: { 'Idempotency-Key': idempotencyKey } },
  )
}

export function withdrawRecruitment(
  recruitId: string,
  idempotencyKey: string,
) {
  return remove<{ status: 'withdrawn'; recruit_id: string }>(
    `/api/recruitments/${recruitId}`,
    { headers: { 'Idempotency-Key': idempotencyKey } },
  )
}
